### PR TITLE
[flang] Set assumed-size last extent to -1

### PR DESCRIPTION
**Pull request by jeanPerier**
Currently lowering sets the extents of assumed-size array to "undef" which was OK as long as the value was not expected to be read.

But when interfacing with the runtime and when passing assumed-size to assumed-rank, this last extent may be read and must be -1 as specified in the BIND(C) case in 18.5.3 point 5.

Set this value to -1, and update all the lowering code that was looking for an undef defining op to identify assumed-size: much safer to propagate and use semantic info here, the previous check actually did not work if the array was used in an internal procedure (defining op not visible anymore).

@clementval and @agozillon, I left assumed-size extent to zero in the acc/omp bounds op as it was, please double check that is what you want (I can imagine -1 may create troubles here, and 0 makes some sense as it would lead to no data transfer).

This also allows removing special cases in UBOUND/LBOUND lowering.

Also disable allocation of cray pointee. This was never intended and would now lead to crashes with the -1 value for assumed-size cray pointee.

### CODE REVIEW

File: flang/lib/Optimizer/Builder/IntrinsicCall.cpp [hunk 1]
```diff
@@ -5691,10 +5691,10 @@ static mlir::Value computeLBOUND(fir::FirOpBuilder &builder, mlir::Location loc,
   if (hasDefaultLowerBound(array))
     return one;
   mlir::Value lb = fir::factory::readLowerBound(builder, loc, array, dim, one);
-  if (dim + 1 == array.rank() && array.isAssumedSize())
-    return lb;
   mlir::Value extent = fir::factory::readExtent(builder, loc, array, dim);
   zero = builder.createConvert(loc, extent.getType(), zero);
+  // Note: for assumed size, the extent is -1, and the lower bound should
+  // be return. It is important to test extent == 0 and not extent > 0.
```
Comment (clementval) [inline flang/lib/Optimizer/Builder/IntrinsicCall.cpp:5697]:
```suggestion
  // be returned. It is important to test extent == 0 and not extent > 0.
```

File: flang/lib/Optimizer/Builder/IntrinsicCall.cpp [hunk 2]
```diff
@@ -5703,52 +5703,29 @@ static mlir::Value computeLBOUND(fir::FirOpBuilder &builder, mlir::Location loc,
...
+        mlir::Value localShape = builder.createShape(loc, array);
+        mlir::Value oldBox = boxValue.getAddr();
+        return builder.create<fir::ReboxOp>(loc, oldBox.getType(), oldBox,
+                                            localShape,
+                                            /*slice=*/mlir::Value{});
       },
-      [&](const auto &) -> fir::ExtendedValue {
-        fir::emitFatalError(loc, "not an assumed size array");
+      [&](const auto &) -> mlir::Value {
+        // This a pointer/allocatable, or an entity not yet tracked with a
```
Comment (clementval) [inline flang/lib/Optimizer/Builder/IntrinsicCall.cpp:5722]:
```suggestion
        // This is a pointer/allocatable, or an entity not yet tracked with a
```

File: flang/lib/Optimizer/Transforms/ArrayValueCopy.cpp [hunk 3]
```diff
@@ -822,6 +821,16 @@ static mlir::Type getEleTy(mlir::Type ty) {
   return ReferenceType::get(eleTy);
 }
 
+// This is an unsafe safe way to deduce this (won't be true in internal
```
Comment (clementval) [inline flang/lib/Optimizer/Transforms/ArrayValueCopy.cpp:824]: unsafe safe? Is this what you meant?

**Review by clementval — APPROVED** [review]
LGTM. I think we are ok to keep 0 in genBounds at least for the time being.

File: flang/lib/Lower/ConvertVariable.cpp [hunk 4]
```diff
@@ -1485,7 +1501,8 @@ lowerExplicitExtents(Fortran::lower::AbstractConverter &converter,
             computeExtent(builder, loc, lowerBounds[spec.index()], ub));
     } else if (spec.value()->ubound().isStar()) {
       // Assumed extent is undefined. Must be provided by user's code.
-      result.emplace_back(builder.create<fir::UndefOp>(loc, idxTy));
+      // result.emplace_back(builder.create<fir::UndefOp>(loc, idxTy));
```
Comment (vzakhari) [inline flang/lib/Lower/ConvertVariable.cpp:1504]: Please remove.

**Review by vzakhari — APPROVED** [review]

File: flang/lib/Lower/ConvertVariable.cpp (hunk 4 above)
Comment (jeanPerier) [reply flang/lib/Lower/ConvertVariable.cpp]: Removed and updated outdated comments in this file, thanks.

**Comment by jeanPerier** [comment]
> LGTM. I think we are ok to keep 0 in genBounds at least for the time being.

Thanks for the review!

**Pull request merged**